pkg/operator/operator.go

```diff
@@ -743,3 +776,244 @@ func (o *Operator) Config(ctx context.Context, key string) (*manifests.Config, e
 
 	return c, nil
 }
+
+// Upgradeable verifies whether the operator can be upgraded or not. It returns
+// the ConditionStatus with optional reason and message.
+func (o *Operator) Upgradeable(ctx context.Context) (configv1.ConditionStatus, string, string, error) {
+	if !o.lastKnowInfrastructureConfig.HighlyAvailableInfrastructure() {
+		return configv1.ConditionTrue, "", "", nil
+	}
+
+	workloadsCorrectlySpread, reason, message, err := o.WorkloadsCorrectlySpread(ctx)
+	if err != nil {
+		return configv1.ConditionUnknown, "", "", err
+	}
+
+	if !workloadsCorrectlySpread {
+		return configv1.ConditionFalse, reason, message, nil
+	}
+
+	return configv1.ConditionTrue, reason, message, nil
+}
+
+func (o *Operator) ensureNodesAreUncordonned(ctx context.Context) error {
+	for nodeName := range o.nodesCordon {
+		node, err := o.client.GetNode(ctx, nodeName)
+		if apierrors.IsNotFound(err) {
+			delete(o.nodesCordon, nodeName)
+			continue
+		}
+		if err != nil {
+			return err
+		}
+
+		err = drain.RunCordonOrUncordon(o.drainer, node, false)
+		if err != nil {
+			return err
+		}
+		delete(o.nodesCordon, node.Name)
+	}
+	return nil
+}
+
+func (o *Operator) spreadWorkloads(ctx context.Context, namespace string, sel map[string]string) (bool, error) {
+	podList, err := o.client.ListPods(ctx, namespace, metav1.ListOptions{LabelSelector: labels.FormatLabels(sel)})
+	if err != nil {
+		return false, err
+	}
+
+	if len(podList.Items) <= 1 {
+		return false, nil
+	}
+
+	node, err := o.client.GetNode(ctx, podList.Items[0].Spec.NodeName)
+	if err != nil {
+		return false, err
+	}
+
+	err = drain.RunCordonOrUncordon(o.drainer, node, true)
```

Comment:
I would cordon the node at the very last moment (e.g. when we're ready to delete the PVC and pod).